add rt1308 and rt711 support

bardliao opened the pull request:
Add support for rt1308 and rt711. Tested on ICL with Realtek ALC711 AIOC TEST CARD.
We may want to use a better way to get the dai index, and "ASoC: SOF: intel: get dai index from dai name" can be a temporary solution.

bardliao committed:
ASoC: SOF: intel: get dai index from dai name

The dai index comes from topology. However, we don't have a good
mechanism to get the topology data in the hw_params or hw_free function.
The patch assume that there is only one BE dai on each soundwire bus.
And we should use the ALH port with the same index as soundwire bus.
e.g.
We should use ALH 1 if the codec is on soundwire bus 1.

Signed-off-by: Bard Liao <yung-chuan.liao@linux.intel.com>

bardliao committed:
soundwire: set grp_ctrl_valid false

Some Realtek codecs don't support grp_ctrl_valid. Set it to false
to prevent DPN_BlockCtrl2 reg write failed.

Signed-off-by: Bard Liao <bardliao@gmail.com>

bardliao committed:
ASoC: rt1308: add soundwire support

Signed-off-by: Bard Liao <bardliao@gmail.com>

bardliao committed:
soundwire: intel: remove playback/capture stream_name

We will create dai widget in SOF.

Signed-off-by: Bard Liao <yung-chuan.liao@linux.intel.com>

plbossart commented on sound/soc/intel/boards/cnl_rt700.c:
```diff
@@ -105,7 +105,8 @@ static const struct snd_soc_dapm_widget cnl_rt700_widgets[] = {
 	{ "Headphones", NULL, "HP" },
-	{ "Speaker", NULL, "SPK" },
+	{ "Speaker", NULL, "SPOL" },
+	{ "Speaker", NULL, "SPOR" },
```
I think this should be a different route (in a different structure), e.g. we need to keep the ability to test ALC700 only for on-board solutions, without RT1308. So I'd rather have a quirk and different routes really.

plbossart commented on sound/soc/intel/boards/cnl_rt700.c:
```diff
@@ -148,6 +149,11 @@ SND_SOC_DAILINK_DEF(sdw0_pin,
+	DAILINK_COMP_ARRAY(COMP_CPU("SDW2 Pin0")));
+SND_SOC_DAILINK_DEF(sdw2_codec,
+	DAILINK_COMP_ARRAY(COMP_CODEC("sdw:2:25d:1308:0:0", "rt1308-aif")));
+
```
lets add sdw3 as well?

bardliao replied:
@plbossart We don't have rt715 codec driver for now. So, the sound card will not probed if we add sdw3.

plbossart commented on sound/soc/intel/boards/cnl_rt700.c:
```diff
@@ -178,6 +184,7 @@ SND_SOC_DAILINK_DEF(platform,
 		DAILINK_COMP_ARRAY(COMP_PLATFORM("0000:00:1f.3")));
 
 struct snd_soc_dai_link cnl_rt700_msic_dailink[] = {
+#if 1
```
we should have different dailinks depending on whether we use the AOB or not.

plbossart commented on sound/soc/intel/boards/cnl_rt700.c:
```diff
@@ -250,6 +248,13 @@ struct snd_soc_dai_link cnl_rt700_msic_dailink[] = {
+static struct snd_soc_codec_conf rt1308_codec_conf[] = {
+	{
+		.dev_name = "sdw:2:25d:1308:0:0",
+		.name_prefix = "rt1308",
```
the prefix is not going to scale if you have 2 rt1308, so need to have rt1308-1 or something

plbossart commented on sound/soc/codecs/rt711-sdw.c:
```diff
@@ -0,0 +1,334 @@
+	/* set the timeout values */
+	prop->clk_stop_timeout = 20;
+
+	return 0;
```
this looks the same as the rt700 code? Is there any way we can factor common things between RT700 and RT711?

plbossart commented on sound/soc/codecs/rt711-sdw.c:
```diff
@@ -0,0 +1,334 @@
+}
+
+module_init(sdw_slave_init);
+module_exit(sdw_slave_exit);
```
same comments, use sdw_module_driver.

plbossart commented on sound/soc/intel/boards/cnl_rt700.c:
```diff
@@ -149,7 +149,7 @@ static int cnl_rt700_codec_fixup(struct snd_soc_pcm_runtime *rtd,
 	DAILINK_COMP_ARRAY(COMP_CPU("SDW0 Pin0")));
 SND_SOC_DAILINK_DEF(sdw0_codec,
-	DAILINK_COMP_ARRAY(COMP_CODEC("sdw:0:25d:700:0:0", "rt700-aif1")));
+	DAILINK_COMP_ARRAY(COMP_CODEC("sdw:0:25d:711:0:1", "rt711-aif1")));
```
this should be added and used inside a different dailink when we use the AOC

plbossart commented on sound/soc/codecs/rt1308-sdw.c:
```diff
@@ -0,0 +1,779 @@
+		init_completion(&slave->port_ready[i]);
+
+	/* set the timeout values */
+	prop->clk_stop_timeout = 20;
```
same copy/paste it seems. maybe we can have a common function between all rtk codecs for inits.

plbossart reviewed (changes requested):
Thanks @bardliao I think the main issue is that we want to keep functionality on the default ICL boards, as well as support the AOC board, so I think we need to have different quirks and dailinks rather than a configuration that only supports AOC.
For the dai index and grouping will look at this separately. Thanks for providing the pointers.

bardliao committed:
ASoC: rt711: add rt711 support

Signed-off-by: Bard Liao <yung-chuan.liao@linux.intel.com>

bardliao committed:
ASoC: intel: cnl_rt700: add rt1308 and rt711 support

Add support for rt711 and rt1308. Use is_using_rt711 to config
if rt711 + rt1308 are using.

Signed-off-by: Bard Liao <bardliao@gmail.com>

bardliao commented:
> Thanks @bardliao I think the main issue is that we want to keep functionality on the default ICL boards, as well as support the AOC board, so I think we need to have different quirks and dailinks rather than a configuration that only supports AOC.

@plbossart I squashed some patches and use module param as a quirk to define if AOC is used. I didn't touch codec driver part. Maybe @shumingfan can chime in?
